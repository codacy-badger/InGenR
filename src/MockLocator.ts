import { GeneratorLocator, Generator } from "./GeneratorLocator"
import _debug from "debug"

const debug = _debug("InGenR:GeneratorLocator")

export class MockLocator extends GeneratorLocator {
  public mockGenerators = new Map<string, Generator | { default: Generator }>()
  public generatorDir = "<project-root>/ingenr-generators"
  // tslint:disable-next-line:no-empty
  async bootstrap() {}
  protected requireModule(requirable: string) {
    debug("Attempting to require mocked module", requirable)
    const retrieved: any = this.mockGenerators.get(requirable)
    if (retrieved) {
      return retrieved.default || retrieved
    }
    return retrieved
  }
}
