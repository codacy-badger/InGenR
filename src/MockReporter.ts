import { Reporter, ConsoleReporter } from "./ConsoleReporter"

export class MockReporter extends ConsoleReporter {
  // tslint:disable-next-line:no-empty
  reportAllWarnings() {}
}
