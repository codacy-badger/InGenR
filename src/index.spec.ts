import * as fs from "fs-extra"
import * as path from "path"
import * as tmp from "tmp"
import _debug from "debug"
import { run, processProject } from "."
import { MockReporter } from "./MockReporter"
import { MockLocator } from "./MockLocator"

const debug = _debug("InGenR:spec:index")

const populateFixtures = async (projDir: string, filePaths: string[]) => {
  for (const filePath of filePaths) {
    const srcPath = path.join(__dirname, "__fixtures__", filePath)
    const destPath = path.join(projDir, filePath)
    await fs.ensureDir(path.dirname(destPath))
    const content = await fs.readFile(srcPath, { encoding: "utf8" })
    await fs.writeFile(destPath, content)
  }
}

const getTmpDir = () =>
  new Promise<string>((resolve, reject) => {
    tmp.dir({ keep: true }, (err, dir) => {
      if (err) reject(err)
      else resolve(dir)
    })
  })

describe("InGenR", () => {
  let projDir: string | null = null
  let originalCwd: string | null = null

  const readFromProjDir = (filePath: string) =>
    fs.readFile(path.join(projDir!, filePath), {
      encoding: "utf8"
    })

  beforeEach(async () => {
    originalCwd = process.cwd()
    projDir = await getTmpDir()
    process.chdir(projDir)
  })
  afterEach(async () => {
    process.chdir(originalCwd!)
    if (projDir) {
      fs.remove(projDir).catch(e => {
        console.error(`Failed to remove temporary directory created for tests: ${projDir}`)
        debug("Error: ", e)
      })
      projDir = null
    }
  })
  it("Injects generated content into annotated blocks", async () => {
    await populateFixtures(projDir!, ["src/index.ts", "ingenr-generators/knex-dal.dot"])
    await run()
    expect(await readFromProjDir("src/index.ts")).toMatchSnapshot()
  })
  it("Complains about missing generators", async () => {
    const srcFiles = ["src/index.ts", "src/missing-generator.ts", "src/erroneous-template-name.ts"]
    await populateFixtures(projDir!, [...srcFiles, "ingenr-generators/knex-dal.dot"])
    const reporter = new MockReporter()
    await processProject(srcFiles.map(p => path.join(projDir!, p)), undefined, undefined, reporter)
    expect(reporter.stripWarnings(projDir!)).toMatchSnapshot()
  })
  it("supports external template targets", async () => {
    await populateFixtures(projDir!, [
      "src/index.ts",
      "src/external-target.ts",
      "ingenr-generators/knex-dal.dot"
    ])
    await run()
    const postProcessedContents = {
      index: await readFromProjDir("src/index.ts"),
      externalTarget: await readFromProjDir("src/external-target.ts"),
      externalTargetGenerated: await readFromProjDir("src/users-table.ts")
    }
    expect(postProcessedContents).toMatchSnapshot()
  })
  it("supports multiple targets and generator modules", async () => {
    const srcFiles = ["src/multiple-targets.ts"]
    await populateFixtures(projDir!, srcFiles)
    const mockReporter = new MockReporter()
    const mockLocator = new MockLocator({
      reporter: mockReporter,
      generatorsDir: true
    })
    mockLocator.mockGenerators.set("pg-dal", {
      default: () => `hello world`
    })
    mockLocator.mockGenerators.set(
      path.join(projDir!, "ingenr-generators", "sqlite-dal"),
      () => `lorem ipsum`
    )
    await processProject(
      srcFiles.map(p => path.join(projDir!, p)),
      undefined,
      mockLocator,
      mockReporter
    )
    const snapshot = {
      src: await readFromProjDir("src/multiple-targets.ts"),
      warnings: mockReporter.stripWarnings(projDir!)
    }
    expect(snapshot).toMatchSnapshot()
  })
})
