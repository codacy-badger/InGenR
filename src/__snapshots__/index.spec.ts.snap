// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InGenR Complains about missing generators 1`] = `
Array [
  Object {
    "filePath": "<project-root>/src/erroneous-template-name.ts",
    "line": undefined,
    "lineIndex": undefined,
    "warnings": Array [
      Object {
        "message": "Invalid template name: knex/../../../foo-bar",
      },
    ],
  },
  Object {
    "filePath": "<project-root>/src/missing-generator.ts",
    "line": undefined,
    "lineIndex": undefined,
    "warnings": Array [
      Object {
        "message": "Failed to resolve generator: knex-dahl",
      },
    ],
  },
]
`;

exports[`InGenR Complains about missing generators 2`] = `
"import foo from \\"foo\\"

const bar = \\"bar\\"

/*! InGenR:expand knex-dal
 * ---
 * tableName: users
 * columns:
 *   - name: name
 *     type: string
 *   - name: email
 *     type: string
 */
interface users {
  name: string;
  email: string;
  
}

const createTable = () =>
  knex.schema.createTable(\\"users\\", (table) => {
      table.uuid(\\"id\\").primary();
      table.string(\\"name\\")
      table.string(\\"email\\")
      
  })
/*! InGenR:end */

export { foo, bar }
"
`;

exports[`InGenR Injects generated content into annotated blocks 1`] = `
"import foo from \\"foo\\"

const bar = \\"bar\\"

/*! InGenR:expand knex-dal
 * ---
 * tableName: users
 * columns:
 *   - name: name
 *     type: string
 *   - name: email
 *     type: string
 */
interface users {
  name: string;
  email: string;
  
}

const createTable = () =>
  knex.schema.createTable(\\"users\\", (table) => {
      table.uuid(\\"id\\").primary();
      table.string(\\"name\\")
      table.string(\\"email\\")
      
  })
/*! InGenR:end */

export { foo, bar }
"
`;

exports[`InGenR supports external template targets 1`] = `
"import foo from \\"foo\\"

const bar = \\"bar\\"

/*! InGenR:expand knex-dal
 * ---
 * tableName: users
 * columns:
 *   - name: name
 *     type: string
 *   - name: email
 *     type: string
 */
interface users {
  name: string;
  email: string;
  
}

const createTable = () =>
  knex.schema.createTable(\\"users\\", (table) => {
      table.uuid(\\"id\\").primary();
      table.string(\\"name\\")
      table.string(\\"email\\")
      
  })
/*! InGenR:end */

export { foo, bar }
"
`;

exports[`InGenR supports external template targets 2`] = `
"import foo from \\"foo\\"

const bar = \\"bar\\"

/*! InGenR:expand knex-dal
 *
 * targetFilePath: users-table.ts
 * ---
 * tableName: users
 * columns:
 *   - name: name
 *     type: string
 *   - name: email
 *     type: string
 */
/*! InGenR:end */

export { foo, bar }
"
`;

exports[`InGenR supports external template targets 3`] = `
"interface users {
  name: string;
  email: string;
  
}

const createTable = () =>
  knex.schema.createTable(\\"users\\", (table) => {
      table.uuid(\\"id\\").primary();
      table.string(\\"name\\")
      table.string(\\"email\\")
      
  })
"
`;

exports[`InGenR supports multiple targets and generator modules 1`] = `
"import foo from \\"foo\\"

const bar = \\"bar\\"

/*! InGenR:expand pg-dal,sqlite-dal
 * ---
 * tableName: users
 * columns:
 *   - name: name
 *     type: string
 *   - name: email
 *     type: string
 */
hello world
lorem ipsum
/*! InGenR:end */

export { foo, bar }
"
`;

exports[`InGenR supports multiple targets and generator modules 2`] = `Array []`;
